src/main/java/com/google/devtools/build/lib/analysis/LocationExpander.java

```diff
@@ -362,9 +362,22 @@ private String functionName() {
       }
     }
 
+    boolean inAspect = ruleContext.getAspects().size() > 0;
+
     // Add all destination locations.
     for (OutputFile out : ruleContext.getRule().getOutputFiles()) {
-      mapGet(locationMap, out.getLabel()).add(ruleContext.createOutputArtifact(out));
+      Artifact artifact;
+      if (inAspect) {
+        // In aspect processing, so don't create & verify a new artifact, just find any cached ones.
+        artifact = ruleContext.getOutputArtifactsProvider().getOutputArtifact(out);
+        // FIXME: I see correct output if I just ignore missing artifacts here. Is that problematic?
+        if (artifact != null) {
+          mapGet(locationMap, out.getLabel()).add(artifact);
```

Comment:
I've been thinking about this for a while and I couldn't come up with a simple and obviously correct without runtime costs :(

How about cutting the Gordian knot by just... making `ctx.expand_location()` not expand the output files of a rule in aspects attached to it? Such location expansion doesn't work at HEAD (since you'd need to create an action generating it in the aspect and it would be an action conflict, because the rule already creates one) so it would be a backwards compatible change and this functionality is not useful, since aspects don't get access to the `OutputFile` instances of the rule they are attached to in general.

Reply:
How would we handle the particular case of #7083? Namely:

```
objc_library(
    name = "simple",
    srcs = glob(["*.m"]),
    hdrs = glob(["*.h"]),
    copts = ["-I$(location headermap)"],
    deps = ["headermap"],
    visibility = ["//visibility:public"],
)
```

Where Tulsi is running an aspect that takes whatever is in the `copts` attribute and passes it down to Xcode in the `HEADER_SEARCH_PATHS` variable? If the aspect doesn't expand the `$(location)` macro (See https://github.com/bazelbuild/tulsi/issues/74#issuecomment-458217208), then Xcode ends up passing the literal string `-I$(location headermap)` to `clang`.

Or perhaps there is a better way of passing the location of the output of the `headermap` rule to the `copt` attribute of the `objc_library()` rule.

Reply:
That's a different case, if I understand correctly: `$(location headermap)` does not expand to an output file of the `objc_library` rule, but to the `FileProvider` of one of its dependencies, which should be handled by the (ugly, but bug-free) special-casing of the `data`/`deps`/`tools` attributes here:

https://github.com/bazelbuild/bazel/blob/master/src/main/java/com/google/devtools/build/lib/analysis/LocationExpander.java#L379

This bug happens not because the output of the `ruleContext.createOutputArtifact(out)` call is used, but simply because it's called: it being called is enough to create an `Artifact` instance which will then be left without a generating action.

Reply:
Hmm... seems we've been talking about two different things then... in that Issue, the `simple` rule was the `objc_library` one.

> The failure is due to the outputs of the simple rule being added to the list of required outputs, but note that simple is not required to expand the location, it's headermap that is required.

I agree with you on the outputs of the rule itself not being expanded, but somehow we need to remove that call to `ruleContext.createOutputArtifact(out)` or somehow make it *not* track the artifacts. That was my first attempt, but I didn't realize I was turning off artifact tracking for _all_ aspects.

Reply:
So is the suggestion here that, when evaluating aspects, location expansion wouldn't need to call `#createOutputArtifact()` at all? If so, that would seem to be an easy change to make.

Reply:
Yep, that's exactly my suggestion -- just make it so that if `LocationExpander` is called from an aspect, it does not call `ruleContext.getRule().getOutputFiles()` or add its output to the set of known labels. Every other label should be fine, because those don't result in the creation of an `Artifact` instance.

@ob : sorry, I was little confused -- I thought `:headermap` was an `objc_library` rule for some reason.

Reply:
@lberki cool! Seems like we are in agreement... as @AlanQuatermain said, it should be a pretty straightforward change. Thanks for all your feedback and patience.